File: tmp/chemicalspace.py
import sys

class Element(object):
  def __init__(self,name,weight,entropy):
    self.name=    name
    self.weight=  weight
    self.entropy= entropy

class Species(object):
  def __init__(self,name,typ,formula):
    self.name=    name
    self.typ=     typ
    self.formula= formula
    self.list_elem_names,self.list_coeffs= formula_read(formula)

def formula_read(s):
  list_elements= []
  list_coeffs= []
  w1= s.split(')')
  for x in w1:
    if '(' in x:
      w2= x.split('(')
      y= w2[0] # ;  print y
      z= w2[1] # ;  print z
      list_elements.append(y.capitalize())
      list_coeffs.append(z)
  return list_elements, list_coeffs

class PhaseModel(object):
  def __init__(self,name,eos):
    self.name=    name
    self.eos=     eos
    self.species= []
  def add(self, x):
    self.species.append(x)

class Phase(object):
  def __init__(self,name,model,composition):
    self.name= name
    self.model= model
    self.composition= composition
  
class ChemSpace(object):
  def __init__(self,name,elements,species,models):
    self.name= name
    self.elements= elements
    self.species= species
    self.models= models



#----------------------------------------------------------read elements
def elements_readfile(path):
  f=open(path,'r')
  elements= []
  for line in f:
    if line.strip()=='': continue
    if line[0]=="%":     continue
    w= line.split()
    if len(w)>1:
      x= Element(name=w[0].capitalize(),entropy=0.,weight=w[1])
      elements.append(x)
  f.close()
  return elements
#--------------------------------------------------------//read elements

#-----------------------------------------------------------read species
def species_readfile(path):
  f=open(path,'r')
  species= []
  for line in f:
    if line.strip()=='': continue
    if line[0]=="%": continue
    w= line.split()
    if len(w)>2:
      x= Species(typ=w[1],name=w[1],formula=w[2])
      species.append(x)
  f.close()
  return species
#---------------------------------------------------------//read species

#-----------------------------------------------------------reading test
elements= elements_readfile("data/atomic_masses.dat")
for x in elements:
  print x.name, x.weight

species= species_readfile("data/species.dat")
for x in species:
  print x.typ, x.name, x.formula, x.list_elem_names, x.list_coeffs

elements_all= elements
list_elem_all= [x.name for x in elements_all]
elements= []
elem_names= []
for spc in species:
  for ele in spc.list_elem_names:
    if ele in list_elem_all:
      if not ele in elem_names:
        elem_names.append(ele)
        j= list_elem_all.index(ele)
        elements.append(elements_all[j])
print elem_names
for x in elements:
  print x.name, x.weight

water_model= PhaseModel("WATER","DILUTE")
for spc in species:
  if spc.typ=="AQU":
    water_model.species.append(spc)
  
def formula_table_compute(species,elem_names):
  formula_table= []
  for spc in species:
    v= [0 for j in range(len(elem_names))]
    i=0
    for ele in spc.list_elem_names:
      #print ele
      j= elem_names.index(ele)
      v[j]= spc.list_coeffs[i]
      i+=1
    formula_table.append(v)
  return formula_table

formula_table= formula_table_compute(species,elem_names)

print formula_table
#---------------------------------------------------------//reading test

models= []
model= PhaseModel(name="name_1",eos="ideal")
for spc in species:
  model.add(spc)
models.append(model)

model= PhaseModel(name="name_2",eos="ideal")
for spc in species:
  model.add(spc)
models.append(model)

myspace= ChemSpace(
  name="myspace",
  elements= elements,
  species= species,
  models= models)

sys.exit()

File: tmp/chemicalspace.jl
type Element
  name::   String
  weight:: String  #Float32
end

type Species
  name::        String
  typ::         String
  formula::     String
  list_elem::   Vector{String}
  list_coeffs:: Vector{Int8}
  #list_elem::   String[]
  #list_coeffs:: String[]
  #x::Array{Float64,1}
  #dim::         Integer
  #elements::    Array{Element} #[dim]
  #stoichio::    Array{Integer} #[dim]
end

type PhaseModel
  name::    String
  species:: Vector{Species}
  eos::     String
end

type Phase
  name::        String
  model::       PhaseModel
  composition:: Vector{Float64}
end
  
type ChemicalSpace
  name::        String
  elements::    Vector{Element}
  species::     Vector{Species}
  models::      Vector{PhaseModel}
end

function formula_read(s)
  list_elements= String[]
  list_coeffs=   Int8[]
  if contains(s,")")
    w1= split(s,")")
    for x in w1
      if contains(x,"(")
        w2= split(x,"(")
        y= ucfirst(lowercase(w2[1])) # ;  print y
        z= parse(Int8,w2[2]) # ;  print z
        push!(list_elements, y)    # list_elements.append(y.capitalize())
        push!(list_coeffs,   z)    # list_coeffs.append(z)
      end
      #println(list_elements)
    end
  end
  return list_elements, list_coeffs
end

if false
s= "AL(1)O(1)H(1)+(2)"
list_elements, list_coeffs= formula_read(s)
println(list_elements)
end

#-----------------------------------------------------------read species
function species_readfile(path)
  f=open(path)
  lines = readlines(f)
  close(f)

  species= Species[]
  for line in lines
    if strip(line)==""        ; continue ; end
    if startswith(line,"%")   ; continue ; end
    w= split(line)
    if length(w)>2
      list_elements, list_coeffs= formula_read(w[3])
      x= Species(w[2],w[1],w[3],list_elements, list_coeffs)
      push!(species, x)
    end
  end
  
  return species
end
#---------------------------------------------------------//read species

#----------------------------------------------------------read elements
function elements_readfile(path)
  f=open(path)
  lines = readlines(f)
  close(f)
  
  elements= Element[]
  for line in lines
    if strip(line)==""        ; continue ; end
    if startswith(line,"%")   ; continue ; end
    w= split(line)
    if length(w)>1
      x= Element(ucfirst(lowercase(w[1])),w[2])
      push!(elements, x)
    end
  end
  
  return elements
end
#--------------------------------------------------------//read elements

#-----------------------------------------------------------reading test
elements= elements_readfile("data/atomic_masses.dat")
if false
  for x in elements
    println(x.name," ",x.weight)
  end
end

species= species_readfile("data/species.dat")
if false
  for x in species
    println( x.typ," ", x.name," ", x.formula ) #, x.list_elem_names, x.list_coeffs )
  end
end
if false
  for x in species
    println(x.list_coeffs )
  end
end

elements_all= elements
list_elem_all= [x.name for x in elements_all]
elements= Element[]
elem_names= String[]
for spc in species
  for ele in spc.list_elem
    #println(ele)
    if ele in list_elem_all
      if !(ele in elem_names)
        push!(elem_names, ele)
        j= findfirst(list_elem_all,ele)
        push!(elements, elements_all[j])
      end
    end
  end
end

#println(elem_names)
for x in elements
  println(x.name," ",x.weight)
end

formula_table= [0 for i in 1:length(species), c in 1:length(elements)]
#formula_table= zeros(length(species),length(elements))
i=1
for spc in species
  j=1
  for ele in spc.list_elem
    k= findfirst(elem_names,ele)
    formula_table[i,k]= spc.list_coeffs[j]
    j+=1
  end
  i+=1
end

for i in 1:length(species)
  println(formula_table[i,:]," ",species[i].name)
end

#println(formula_table)
#---------------------------------------------------------//reading test

mixmodels= PhaseModel[]
model= PhaseModel("name_1",species,"ideal")
push!(mixmodels,model)
model= PhaseModel("name_2",species,"ideal")
push!(mixmodels,model)

myspace= ChemicalSpace("myspace", elements,species,mixmodels)

quit()


